sharktank/tests/kernels/attention_template_test.py

```diff
@@ -0,0 +1,138 @@
+
+import unittest
+from parameterized import parameterized
+
+import torch
+
+from iree.turbine import aot
+from sharktank import kernels
+from sharktank.types import layout_utils
+from sharktank.utils import debugging
+from sharktank import ops
+
+
+class custom_attention(unittest.TestCase):
+    def setUp(self):
+        torch.manual_seed(420)
+
+    @parameterized.expand(
+        [
+            (torch.float32, 5e-3, 1e-3, True),
+            (torch.float16, 5e-3, 1e-3, True),
+            # Currently failing on unmasked error
+            # (torch.float32, 5e-3, 1e-3, False),
+            # (torch.float16, 5e-3, 1e-3, False),
+        ]
+    )
+    def test_compare_torch_spda(self, dtype, atol, rtol, use_mask):
+        H = 4  # Head dim
+        N = 3  # Batch Size
+        L = 7  # Target Seq Len
+        S = 6  # Source Seq Len
+        Eqk = Ev = 64  # embedding dimensions with subscript identifiers
+
+        q = torch.rand([N, H, L, Eqk], dtype=dtype)
+        k = torch.rand([N, H, S, Eqk], dtype=dtype)
+        v = torch.rand([N, H, S, Ev], dtype=dtype)
+        # mask is same type as inputs, therefore its added to score
+        mask = None
+        scale = torch.tensor(1.0, dtype=dtype)
+        if use_mask:
+            mask = torch.rand([L, S], dtype=dtype)
+
+            res2 = kernels.masked_flash_attention(q, k, v, mask, scale=scale)
+
+        else:
+            res2 = kernels.flash_attention(q, k, v, scale)
+
+        ref = torch.nn.functional.scaled_dot_product_attention(
+            q, k, v, mask, scale=scale
+        )
+
+        torch.testing.assert_close(res2.to(dtype), ref, atol=atol, rtol=rtol)
+
+    @parameterized.expand(
+        [
+            # Todo: fixed unmasked.
+            # (torch.float32, False, False),
+            (torch.float32, False, True),
+            (torch.float16, True, True),
+            (torch.float8_e4m3fnuz, False, True),
+        ]
+    )
+    def test_export_dynamic(self, dtype, static, use_mask):
+        ops.attention_impls.register_attention_override_by_name(
+            "masked_flash_attention"
+        )
+        cast = False
+        if dtype == torch.float8_e4m3fnuz:
+            dtype = torch.float32
+            cast = True
+        H = 4  # Head dim
+        N = 3  # Batch Size
+        L = 19  # Target Seq Len
+        S = 19  # Source Seq Len
+        Eqk = Ev = 64  # embedding dimensions with subscript identifiers
+
+        q = torch.rand([N, H, L, Eqk], dtype=dtype)
+        k = torch.rand([N, H, S, Eqk], dtype=dtype)
+        v = torch.rand([N, H, S, Ev], dtype=dtype)
+        if use_mask:
+            # mask is same type as inputs, therefore its added to score
+            mask = torch.rand([L, S], dtype=dtype)
+        if cast:
+            q = q.to(torch.float8_e4m3fnuz)
+            k = q.to(torch.float8_e4m3fnuz)
+            v = v.to(torch.float8_e4m3fnuz)
+            if use_mask:
+                mask = mask.to(torch.float8_e4m3fnuz)
```

Comment:
I think in the real world we would never have a `float8_e4m3fnuz` mask as this type does not support infinites.
In the rest of the PR it seems you want the mask to be `f32`.